plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js

```diff
@@ -1,81 +1,354 @@
+      SET
+    ));
+  }
+
+  if ((command.argv != null) && Array.isArray(command.argv)) {
+    memo.push(new Transaction(
+      path.concat(['command', 'shell']),
+      false,
+      SET
+    ));
+
+    // Always cast to string, since the UI cannot handle arrays
+    memo.push(new Transaction(
+      path.concat(['command', 'string']),
+      command.argv.join(' '),
+      SET
+    ));
+  }
+}
+
+function reduceCommandHealthCheck(newState, field, value) {
+  const {exec: {command={}}} = newState;
+
+  switch (field) {
+    case 'shell':
+      // Shell is a meta-field that denotes if we are going to populate
+      // the argument or the shell field. So, if we encounter an opposite
+      // field, we should convert and set-up a placeholder
+      if (value) {
+        command.shell = '';
+        if ((command.argv != null) && Array.isArray(command.argv)) {
+          command.shell = command.argv.join(' ');
+          delete command.argv;
         }
-        if (`healthChecks.${index}.maxConsecutiveFailures` === joinedPath) {
-          this.healthChecks[index].maxConsecutiveFailures = parseInt(value, 10);
+      } else {
+        command.argv = [];
+        if (command.shell != null) {
+          command.argv = command.shell.split(' ');
+          delete command.shell;
         }
       }
+      break;
+
+    case 'string':
+      // By default we are creating `shell`. Only if `argv` exists
+      // we should create an array
+      if (command.argv != null) {
+        command.argv = value.split(' ');
+      } else {
+        command.shell = value;
+      }
+      break;
+  }
+}
+
+function reduceFormCommandHealthCheck(newState, field, value) {
+  const {exec: {command={}}} = newState;
+
+  switch (field) {
+    case 'shell':
+      command.shell = value;
+      break;
+
+    case 'string':
+      command.string = value;
+      break;
+  }
+}
+
+const MultiContainerHealthChecks = {
+  JSONSegmentReducer(state, {path, value}) {
+    const newState = Object.assign({}, state);
+    const [group, field, secondField] = path;
+
+    // If we are assigning the entire group to `null`, we are
+    // effectively disabling the health checks
+    if ((path.length === 0) && (value == null)) {
+      return null;
+    }
+
+    // Format object structure according to protocol switch
+    if (group === 'protocol') {
+      switch (value) {
+        case COMMAND:
+          newState.exec = {
+            command: {}
+          };
+          delete newState.http;
+          delete newState.tcp;
+          break;
+
+        case HTTP:
+          delete newState.exec;
+          newState.http = {};
+          delete newState.tcp;
+          break;
+
+        case TCP:
+          delete newState.exec;
+          delete newState.http;
+          newState.tcp = {};
+          break;
+      }
+
+      return newState;
+    }
+
+    // Assign properties
+    switch (group) {
+      case 'exec':
+        reduceCommandHealthCheck(newState, secondField, value);
```

Comment:
Comment from @nLight:
https://github.com/dcos/dcos-ui/pull/1728#discussion_r94931809
"I'd rather not mutate newState here
then we could even return reduceFormCommandHealthCheck(...) right away"